packages/devtools_app/lib/src/utils.dart

```diff
@@ -1069,3 +1069,225 @@ String prettyTimestamp(
   );
   return DateFormat.Hms().format(timestampDT); // HH:mm:ss
 }
+
+/// A [ChangeNotifier] that holds a list of data.
+///
+/// This class also exposes methods to interact with the data. By default,
+/// listeners are notified whenever the data is modified, but notifying can be
+/// optionally disabled.
+class ListValueNotifier<T> extends ChangeNotifier
+    implements ValueListenable<List<T>> {
+  /// Creates a [ListValueNotifier] that wraps this value [_rawList].
+  ListValueNotifier(List<T> rawList) : _rawList = List<T>.from(rawList) {
+    _currentList = ImmutableList(_rawList);
+  }
+
+  final List<T> _rawList;
+
+  ImmutableList<T> _currentList;
+
+  @override
+  List<T> get value => _currentList;
+  set value(List<T> newValue) {
+    throw Exception(
+      '[value] cannot be set explicitly. Use the methods on ListValueNotifier'
+      ' to interact with [value].',
+    );
+  }
+
+  void add(T item, {bool notify = true}) {
+    _rawList.add(item);
+    _currentList = ImmutableList(_rawList);
+    if (notify) {
+      notifyListeners();
+    }
+  }
+
+  void addAll(List<T> itemList, {bool notify = true}) {
+    _rawList.addAll(itemList);
+    _currentList = ImmutableList(_rawList);
+    if (notify) {
+      notifyListeners();
+    }
+  }
+
+  bool remove(T element, {bool notify = true}) {
```

Comment:
these remove methods are wrong. They will mess up the previous immutable values. Remove these methods as there isn't an easy way to support them efficiently. The right way to do it would be to copy the _rawList object but no point unless you have a use case.